### ISSUE TITLE
New firmware

### ISSUE THREAD

**Issue by ghostwheel** [HappyZ/dpt-tools#89]
Has anyone tried the new firmware? I just got a prompt to update.
What will happen if I upgrade? Will my hacked one be overwritten?

**Comment by junrenshi** [comment]
I tried to update.  After updated, the device shows the error message like  "Unfortunately the Applauncher stops".  The Home button stops response.  Moreover, DPA does not connect to the device.  dpt-tool cannot connect to the device through the wireless: connection refused.

Fortunately, the device can still be connected in the diagnosis mode.  I guess to revive my device I need to remove the dp_extensions I added into the system.  The trouble is that I cannot see the partition in the diagnosis mode.  I will try to figure out.  Any help/guide will be appreciated.

**Comment by HappyZ** [comment]
@ghostwheel Before we know what exactly was changed in the official update, I would not recommend to update immediately.

@junrenshi it probably messed up with the dp_extensions.. and based on what you said it feels like they've also changed the way you wirelessly connect to the device. I'll have time to look at it this weekend. As I'm not sure about your experience, I do not recommend to do stuff in diagnosis mode directly.

Below is a brief way based on the previous firmware update:

If you are experienced, mount partition `/dev/mmcblk0p9` to see files in `/system/` and you'll find `dp_extensions` in `etc`. Deleting related entries may not be sufficient. You may also need to get back and load `data` partition and change file `/data/system/ExtMgr.db`.

In the worst case, if you have access to diagnosis mode, you can mount the mass storage drive `/dev/mmcblk0p16` and on PC put in the old official firmware FwUpdater.pkg from [here](https://github.com/HappyZ/dpt-tools/wiki/PKGs-I-Made). Then just restart the device so it could flash the old firmware.

**Comment by junrenshi** [comment]
Thanks! I mount /dev/mmcblk0p9.  It turns out the dp_extensions directory is clean -- the update erases my changes.  However,  I am not able to find the data partition.  How can I load the  data partition?

**Comment by HappyZ** [comment]
> Thanks! I mount /dev/mmcblk0p9. It turns out the dp_extensions directory is clean -- the update erases my changes. However, I am not able to find the data partition. How can I load the data partition?

You probably don't have access to data partition (can't mount directly) in diagnosis mode.. I think if you fix dp_extension (add your mods back) it will solve the problem.

**Comment by ultraboy** [comment]
I had this problem before, but I forget how did I solve it. It is important to reset before you flash the official firmware.

**Comment by HappyZ** [comment]
As far as I can tell, the new firmware did not stop the hack. However, I do see LOTS of changes in the firmware pkg. Lots of clean ups happened. I'll leave the rest of checks to weekends.

~~But I haven't seen one removing `dp_extensions`. It may be in the APKs.~~

Like @ultraboy said, a factory reset before upgrading may also be helpful.

**Comment by junrenshi** [comment]
> You probably don't have access to data partition (can't mount directly) in diagnosis mode.. I think if you fix dp_extension (add your mods back) it will solve the problem.

Yes, I tried to add my mods back to dp_extensions, and now it works.  Moreover, KOReader also works.

The only problem now is that the DPA does not connect.

**Comment by chengpeng0409** [comment]
@HappyZ Hey I've also updated my system and got the same problem as @junrenshi did. Can I flash in the official package into the device? I no longer needed the root, just wanted to get my device work again. Thx.

**Comment by Antiparadox** [comment]
How do you guys solve this issue?
I also experienced an AppLauncher failure, and it seems that ADB also stops working...

**Comment by Antiparadox** [comment]
I think the problem is with my changes on dp_extensions, is there any way to undo all the changes on the new firmware? Thanks for any tips.

**Comment by junrenshi** [comment]
I managed to fix everything.

An update: my previous report "dpt-tool cannot connect to the device through the wireless" is false.  It was due to a wrong IP address.  Therefore, you can still connect the device by using the dpt-tool and re-flash the device, as long as you can figure out the IP address.

I re-flashed FWUpdater.pkg without the mass storage support.  Now the device connects to DPA normally.

For the AppLanaucher stop problem, I fixed it by:
(1) run "python3 dpt-tools.py -diagnosis" and follow the instructions;
(2) mount /dev/mmcblk0p9
(3) Add back all previous mods to the directory etc/dp_extensions in the mounted disk.

**Comment by HappyZ** [comment]
@junrenshi  thanks for the brief tutorial. Im surprised the old boot img still works.. it's quite annoying that applauncher works its own way and crashes every time on system updates..

**Comment by Antiparadox** [comment]
I might be doing something really stupid but I cannot do "mount /dev/mmcblk0p9"
here's what happened.

>>> Enter the serial port [/dev/tty.usbmodem01]: /dev/tty.usbmodem01
============================
 DPT Tools - Diagnosis Mode
============================
This is diagnosis mode. Type `help` to show this message.
It behaves similarly to regular serial session with less flexibility (cannot use tab, scroll up, quick reverse search, etc.).
This mode intends to automate some complicated procedures.

Supported commands:
    `push-file`         -- transfer file to DPT at 800bps (=100Bps)
    `pull-file`         -- transfer file from DPT
    `backup-bootimg`    -- backup the boot img and download it to local device
    `restore-bootimg`   -- restore the boot img
    `get-su-bin`        -- enable `su` (root) in adb (beta, not well tested)
    `exit`/`quit`       -- leave the tool
    and many unix cmds (do not support less/head)

root #: ls
boot.img.bak
root@FPX-1010:~# mount /dev/mmcblk0p9
mount: can't find /dev/mmcblk0p9 in /etc/fstab
root@FPX-1010:~# 

I will be very appreciated if any of you guys could look into this error.
Thanks a lot for your help!

**Comment by Antiparadox** [comment]
BTW, I successfully invoked diagnostic mode on my device, and the connection seems fine, though.

**Comment by Antiparadox** [comment]
UPdate:
Got'it, My situation is bit different, I have to manually restore the backup boot-img and then evoke su-access to bring back adb shell, which allows me to remove the local cache for App launcher, and then everything works fine

**Comment by sartrism** [comment]
> UPdate:
> Got'it, My situation is bit different, I have to manually restore the backup boot-img and then evoke su-access to bring back adb shell, which allows me to remove the local cache for App launcher, and then everything works fine

Could you explain more? I'm in exactly the same situation. Which boot.img.bak did you use?

**Comment by Antiparadox** [comment]
@sartrism first kick RP1 into diagnostic mode (press and hold both power Botton and home button to turn on), then use DPT-tools , enter python3 dpt-tools.py -diagnosis, then follow the instructions to connect RP1.
Then enter restore-bootimg in the spawned shell, this will ask you to specify the boot-img that you use, the path would be python_api/assets/boot-1.4.01.16100-mod-happyz-181118
then enter get-su-bin when this is completed, when it's all done, reboot and enter ADB advices in your terminal, you will see the attached devices.
Then enter:
adb shell
su
mount -o rw,remount /system
cd /data/system
mv ExtMgr.db ExtMgr.db_bak
mv ExtMgr.db-journal ExtMgr.db-journal_bak
then reboot
Depending on situations this may not work for you, but it works on my device
Hope it helps, though.

**Comment by sartrism** [comment]
> @sartrism first kick RP1 into diagnostic mode (press and hold both power Botton and home button to turn on), then use DPT-tools , enter python3 dpt-tools.py -diagnosis, then follow the instructions to connect RP1.
> Then enter restore-bootimg in the spawned shell, this will ask you to specify the boot-img that you use, the path would be python_api/assets/boot-1.4.01.16100-mod-happyz-181118
> then enter get-su-bin when this is completed, when it's all done, reboot and enter ADB advices in your terminal, you will see the attached devices.
> Then enter:
> adb shell
> su
> mount -o rw,remount /system
> cd /data/system
> mv ExtMgr.db ExtMgr.db_bak
> mv ExtMgr.db-journal ExtMgr.db-journal_bak
> then reboot
> Depending on situations this may not work for you, but it works on my device
> Hope it helps, though.

Thanks! But is your software up to date by doing so? Flashing the legacy boot.img seems overwriting the new firmware so I cannot see TOC for example (and cannot update because the device thinks it is up to date.)

**Comment by Antiparadox** [comment]
@sartrism no it wouldn't affect the update. DPA is just a software on this devices, the official update would not be affected thus.

**Comment by sartrism** [comment]
> @sartrism no it wouldn't affect the update. DPA is just a software on this devices, the official update would not be affected thus.

I think I messed something while solving this issue, and anyway my DPA reader seems outdated (doesn't show any new features like prev/next page or TOC) though I flashed the new firmware. I will figure it out.

P.S. Oops, actually it seems like the software already has been updated. The prev/next button just show up after I move to other pages, but not initially so I got confused.
